cmd/query/app/apiv3/otlp_translator.go

```diff
@@ -0,0 +1,364 @@
+	for res, libMap := range spansByLibrary {
+		rs := &v1.ResourceSpans{
+			Resource: res.resource,
+		}
+		for _, v := range libMap {
+			rs.InstrumentationLibrarySpans = append(rs.InstrumentationLibrarySpans, v)
+		}
+		rss = append(rss, rs)
+	}
+	return rss
+}
+
+type instrumentationLibrary struct {
+	name, version string
+}
+
+// helper type used as a map key
+type resource struct {
+	serviceName string
+	// concatenated string tags to make sure services with the same name but different versions
+	// are held have distinct resource object.
+	tagsHash uint32
+	resource *resourcev1.Resource
+}
+
+func jSpanToOTLP(jSpan *model.Span) (*v1.Span, resource, instrumentationLibrary) {
+	tags := model.KeyValues(jSpan.GetTags())
+	status, ignoreKeys := getSpanStatus(tags)
+	if ignoreKeys == nil {
+		ignoreKeys = map[string]bool{}
+	}
+
+	traceState := getTraceStateFromAttrs(tags)
+	if traceState != "" {
+		ignoreKeys[tracetranslator.TagW3CTraceState] = true
+	}
+
+	s := &v1.Span{
+		TraceId:           uint64ToTraceID(jSpan.TraceID.High, jSpan.TraceID.Low),
+		SpanId:            uint64ToSpanID(uint64(jSpan.SpanID)),
+		ParentSpanId:      uint64ToSpanID(uint64(jSpan.ParentSpanID())),
+		TraceState:        traceState,
+		Name:              jSpan.GetOperationName(),
+		StartTimeUnixNano: uint64(jSpan.GetStartTime().UnixNano()),
+		EndTimeUnixNano:   uint64(jSpan.GetStartTime().Add(jSpan.GetDuration()).UnixNano()),
+		Events:            jLogsToOTLP(jSpan.GetLogs()),
+		Links:             jReferencesToOTLP(jSpan.GetReferences(), jSpan.ParentSpanID()),
+		Status:            status,
+	}
+	if kind, found := jSpan.GetSpanKind(); found {
+		s.Kind = jSpanKindToInternal(kind)
+		ignoreKeys[tracetranslator.TagSpanKind] = true
+	}
+
+	il := instrumentationLibrary{}
+	if libraryName, ok := tags.FindByKey(conventions.InstrumentationLibraryName); ok {
+		il.name = libraryName.GetVStr()
+		ignoreKeys[conventions.InstrumentationLibraryName] = true
+		if libraryVersion, ok := tags.FindByKey(conventions.InstrumentationLibraryVersion); ok {
+			il.version = libraryVersion.GetVStr()
+			ignoreKeys[conventions.InstrumentationLibraryVersion] = true
+		}
+	}
+	// convert to attributes at the end once not needed attrs are removed
+	attrs := jTagsToOTLP(tags, ignoreKeys)
+	s.Attributes = attrs
+
+	res := resource{}
+	if jSpan.GetProcess() != nil {
+		tags := concatStringTags(jSpan.GetProcess().GetTags())
+		fnva := fnv.New32a()
+		fnva.Write([]byte(tags))
+
+		res.serviceName = jSpan.GetProcess().GetServiceName()
+		res.tagsHash = fnva.Sum32()
+		res.resource = jProcessToInternalResource(jSpan.GetProcess())
+	}
+	return s, res, il
+}
+
+// concatStringTags returns key sorted concatenated string tags
+// e.g. keyB=val1,keyA=val2 becomes keyAval2KeyBval1
+func concatStringTags(tags []model.KeyValue) string {
+	keys := make([]string, len(tags))
+	tagMap := make(map[string]string, len(tags))
+	stringTagsLen := 0
+	for i, t := range tags {
+		if t.GetVType() == model.ValueType_STRING {
```

Comment:
could process tags ever have a non-string value?

Reply:
Yes - added in user-provided OTELcol processor.